src/set_theory/zfc/ordinal.lean

```diff
@@ -84,4 +87,145 @@ theorem is_transitive_iff_subset_powerset : x.is_transitive ↔ x ⊆ powerset x
 
 alias is_transitive_iff_subset_powerset ↔ is_transitive.subset_powerset _
 
+/- ### Von Neumann Hierarchy -/
+
+section von_neumann
+variables {o a b : ordinal.{u}}
+
+/-- The von Neumann hierarchy, defined so that
+
+- $V_{\alpha+1}=\mathcal P(V_\alpha)$,
+- $V_\alpha=\bigcup_{\beta<\alpha}V_\beta$ for $\alpha=0$ or limit.
+
+Every set belongs to the von Neumann hierarchy (`exists_mem_von_neumann`). -/
+-- Todo: the 0 case is the same as the limit case, use an induction principle that merges these.
+noncomputable def von_neumann (o : ordinal.{u}) : Set.{u} :=
+o.limit_rec_on ∅ (λ x, powerset)
+  (λ x hx IH, ⋃₀ range.{u u} (λ i : x.out.α, IH (ordinal.typein (<) i) $ ordinal.typein_lt_self _))
+
+@[simp] theorem von_neumann_zero : von_neumann 0 = ∅ :=
+ordinal.limit_rec_on_zero _ _ _
+
+@[simp] theorem von_neumann_succ (o : ordinal) :
+  von_neumann (order.succ o) = powerset (von_neumann o) :=
+ordinal.limit_rec_on_succ _ _ _ _
+
+theorem von_neumann_limit : o.is_limit →
+  von_neumann o = ⋃₀ range.{u u} (λ i : o.out.α, von_neumann (ordinal.typein (<) i)) :=
+ordinal.limit_rec_on_limit _ _ _ _
+
+theorem mem_von_neumann_limit (ho : o.is_limit) : x ∈ von_neumann o ↔ ∃ a < o, x ∈ von_neumann a :=
+begin
+  simp only [von_neumann_limit ho, mem_sUnion, mem_range, set.mem_range, exists_prop,
+    exists_exists_eq_and],
+  split,
+  { rintro ⟨a, ha⟩,
+    exact ⟨_, ordinal.typein_lt_self _, ha⟩ },
+  { rintro ⟨a, ha, hx⟩,
+    refine ⟨ordinal.enum (<) a _, _⟩;
+    simpa }
+end
+
+theorem von_neumann_transitive (o : ordinal) : (von_neumann o).is_transitive :=
+begin
+  induction o using ordinal.limit_rec_on with o ho o ho IH,
+  { rw von_neumann_zero,
+    exact empty_is_transitive },
+  { rw von_neumann_succ,
+    exact ho.powerset },
+  { intros x hx y hy,
+    rw mem_von_neumann_limit ho at hx ⊢,
+    rcases hx with ⟨b, hb, hx⟩,
+    exact ⟨b, hb, (IH b hb).subset_of_mem hx hy⟩ }
+end
+
+/-- The rank of a set, recursively defined as the least ordinal greater than the ranks of all its
+    members. -/
+noncomputable def rank : Set.{u} → ordinal.{u} :=
```

Comment:
I think this is an acceptable definition for now, but see here for a suggestion of how we could generalize the various rank functions in Mathlib.

https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Small.20initial.20segments